sympy/solvers/solveset.py

```diff
@@ -761,11 +820,33 @@ def _solveset(f, symbol, domain, _check=False):
 
     # assign the solvers to use
     solver = lambda f, x, domain=domain: _solveset(f, x, domain)
-    if domain.is_subset(S.Reals):
-        inverter_func = invert_real
-    else:
-        inverter_func = invert_complex
-    inverter = lambda f, rhs, symbol: inverter_func(f, rhs, symbol, domain)
+
+    # check for Modular equations
+    is_mod = S.false
+    if f.has(Mod):
+        # in case symbol is not defined as integer
+        assumption = symbol.assumptions0
+        assumption['integer'] = True
+        t = Dummy('t', **assumption)
+        f = f.xreplace({symbol: t})
+
+        rep_eq = list(f.atoms(Mod))
+        if all([i.is_integer for i in rep_eq[0].args]):
+            is_mod = S.true
+            rep_eq = rep_eq[0]
+            m = rep_eq.args[1]
+            f = f.replace(rep_eq, rep_eq.args[0])
+            symbol = t
+            inverter = lambda f, rhs, mod, symbol: invert_modular(f, rhs, m, symbol, domain)
+        else:
+            f = f.xreplace({t: symbol})
```

Comment:
If the original replacement `f = f.xreplace({symbol: t})` was inside the if-then clause, you would not have to undo it now, in the else clause.